File: module/bigdata-service/bigdata-start/src/main/java/com/huan/bigdata/run/SparkBatchRunner.scala
package com.huan.bigdata.run

import org.apache.spark.sql.SparkSession
import org.springframework.boot.{ApplicationArguments, ApplicationRunner}
import org.springframework.stereotype.Component

/**
 * hdfs dfs -mkdir input
 * hdfs dfs -put src/main/resources/data/people.json input
 *
 * hdfs dfs -get /user/huanyu/input input
 *
 * hdfs dfs -ls /user/huanyu/input
 */
@Component
class SparkBatchRunner extends ApplicationRunner{
    override def run(args: ApplicationArguments): Unit = {
        try {

            //1:input data path
            val hdfsSourcePath = "hdfs://localhost:9000/user/huanyu/input/people.json"

            //2:create SparkSession
            val spark = SparkSession
                .builder()
                .appName("SparkBatchDemo")
                .master("local")
                //.master("yarn")
                // .config("spark.some.config.option", "some-value")
                .getOrCreate()
            //3:read data from hdfs
            val df = spark.read.json(hdfsSourcePath)
            //4.1: Displays the content of the DataFrame to stdout
            df.show()
            df.printSchema()
            df.select("name").show()
            //4.2 Select everybody, but increment the age by 1
            import spark.implicits._
            df.select($"name", $"age" + 1).show()

            df.filter($"age" > 21).show()

            df.groupBy("age").count().show()
            //5 write result to hdfs
            //  val hdfsTargetPath = "hdfs://127.0.0.1:9000/input/result/"
            //bin/hdfs dfs -rm -r /input/people_result.json
            // df.write.mode(SaveMode.Overwrite) json (hdfsTargetPath)

        } catch {
            case ex: Exception => {
                ex.printStackTrace() // 打印到标准err
                System.err.println("exception===>: ...") // 打印到标准err
            }
        }
    }
}

File: module/bigdata-service/bigdata-start/src/main/java/com/huan/bigdata/run/SparkBatchRunner.java
package com.huan.bigdata.run;

import org.springframework.boot.ApplicationArguments;
import org.springframework.boot.ApplicationRunner;
import org.springframework.stereotype.Component;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;

import java.io.Serializable;

import static org.apache.spark.sql.functions.col;

// @Component
public class SparkBatchRunner implements ApplicationRunner {
    public static class Person implements Serializable {
        private String name;
        private long age;
    
        public String getName() {
          return name;
        }
    
        public void setName(String name) {
          this.name = name;
        }
    
        public long getAge() {
          return age;
        }
    
        public void setAge(long age) {
          this.age = age;
        }
    }

    @Override
    public void run(ApplicationArguments args) throws Exception {
        String hdfsSourcePath = "hdfs://localhost:9000/user/huanyu/input/people.json";
        SparkSession spark = SparkSession
                                .builder()
                                .master("local")
                                .appName("SparkBatchRunner")
                                .getOrCreate();
        Dataset<Row> df = spark.read().json(hdfsSourcePath);
        df.show();
        df.printSchema();
        df.select("name").show();
        df.select(col("name"), col("age").plus(1)).show();
        df.filter(col("age").gt(21)).show();
    }
    
}
